### PR TITLE
[feat] HybridStack grouped syntax and EP-overlap (2/4 of #4798)

### PR DESCRIPTION
**Pull request by Connor-XY**
## What does this PR do?

Part 2 of the original four-part split of #4798 by @Wohox, @Connor-XY, and
@guihong-nv. Adds bracketed HybridStack groups and their EP-overlap scheduling.

Groups such as `[*-]` and `[*E]` form one logical layer containing multiple
physical layers. Construction, recomputation, inference cache indexing, and
checkpoint naming preserve that distinction. Groups reject nested brackets,
MoE members before the final position, and colliding checkpoint namespaces.

The hybrid scheduler supports attention, MLA, Mamba, and MoE callables; uses
the current low-precision context hook; rejects unsupported nested CP-layout
conversions; and preserves NCCLEP probability boundaries, zero-copy staging,
delayed-wgrad hook ownership, and per-microbatch norm-offload managers.

## Checkpoint scope

Grouped stacks use logical GPT-style storage keys. Existing ungrouped Hybrid
checkpoint names and direct-stack offsets remain compatible. Native Hybrid
checkpoint loading stays in scope.

GPT-to-grouped model and optimizer checkpoint import is extracted into a
dependent checkpoint follow-up, including source-format detection and FSDP
mapping. This PR alone does not enable that conversion through the training
loader. The checkpoint follow-up is #7192.

## Review scope

Six CODEOWNERS teams: core-adlr, core-nemo, hybrid-model, gpt,
mixture-of-experts-adlr, and mixture-of-experts-devtech. The checkpoint
follow-up carries the distributed-checkpointing and training-loader files
together with their interoperability tests.

## Dependencies and validation

#4941, the common scheduler refactor, has merged.

Integrated four-GB200 validation at `e8db9ddcb`: 224 cases passed, including
grouped attention/MLA output and gradient parity over three microbatches and
the checkpoint follow-up's tests. Eleven existing Mamba/GDN tests failed
because the cached ARM image contains incompatible native extensions.
The split preserves the exact combined source tree; the remaining Hybrid
regressions and tests stay in this PR.

### CODE REVIEW

**Comment by Connor-XY** [comment]
/ok to test [4aa1be7](https://github.com/NVIDIA/Megatron-LM/pull/4942/commits/4aa1be7c95b63dab68d1c76c5493bb65637c81d3)

File: megatron/core/models/hybrid/hybrid_model.py [hunk 1]
```diff
@@ -489,54 +414,114 @@ def forward(
...
+        attention_mask=None,
+        inference_params=None,
+        packed_seq_params=None,
+        sequence_len_offset=None,
+        runtime_gather_output=None,
+        extra_block_kwargs=None,
+        inference_context=None,
+        is_spec_decode=None,
+        output_processor=None,
+        output_processor_context=None,
```
[the remark this answers is not in the document]
Comment (Connor-XY) [reply megatron/core/models/hybrid/hybrid_model.py:501]: Fixed in 08e141e. `_postprocess` now has the same `if output_processor is not None:` early return as `GPTModel._postprocess`, forwarding the same kwargs (`context=output_processor_context`, `compute_language_model_loss`, `scale_logits`, `config`), placed at the same point — after the MTP block, before the `materialize_only_last_token_logits` slicing.

Left `HybridModel.forward` alone; it doesn't take an output processor today, and `PostProcessNode` is the path that actually forwards one.

New test: `test_hybrid_postprocess_uses_output_processor_hook`.

[the remark this answers is not in the document]
Comment (Connor-XY) [reply megatron/core/models/hybrid/hybrid_model.py:501]: Fixed in 08e141e. `_postprocess` now has the same `if output_processor is not None:` early return as `GPTModel._postprocess`, forwarding the same kwargs (`context=output_processor_context`, `compute_language_model_loss`, `scale_logits`, `config`), placed at the same point — after the MTP block, before the `materialize_only_last_token_logits` slicing.

Left `HybridModel.forward` alone; it doesn't take an output processor today, and `PostProcessNode` is the path that actually forwards one.

New test: `test_hybrid_postprocess_uses_output_processor_hook`.

File: megatron/core/models/hybrid/hybrid_block.py [hunk 2]
```diff
@@ -411,15 +525,15 @@ def sharded_state_dict(
...
-                        metadata,
-                        tp_group=self.tp_group,
-                    )
+                module_prefix = f'{prefix}{name}.'
+                module_sharded_state_dict = sharded_state_dict_default(
+                    module, module_prefix, sharded_offsets, metadata, tp_group=self.tp_group
                 )
+                if name == 'final_norm':
+                    replace_prefix_for_sharding(
+                        module_sharded_state_dict, module_prefix, f'{prefix}final_layernorm.'
```
[the remark this answers is not in the document]
Comment (Connor-XY) [reply megatron/core/models/hybrid/hybrid_block.py]: You're right that this was too broad, and it breaks more than old checkpoints. `docs/user-guide/hybrid-model-migration.md` documents `decoder.final_norm` as *the* HybridModel key, and both `tools/checkpoint/gpt_hybrid_conversion.py` and `tools/checkpoint/remap_gpt_dsa_to_mamba.py` rewrite GPT's `decoder.final_layernorm.*` into `decoder.final_norm.*` when converting. With the rename unconditional, every checkpoint those tools produce would have failed to load.

Now gated, in 08e141e: `HybridStack` takes a `transformer_sharded_keys` flag, and `HybridModel` sets it from whether the *full* layer pattern contains a bracketed group — full pattern rather than this rank's PP segment, so every stage agrees on the naming. Grouped patterns get `final_layernorm` (their logical layers are 1:1 with transformer layers, which is the whole point of the GPT cross-load); everything else keeps `final_norm`, so pre-existing hybrid checkpoints and converter output still load. The converter itself only accepts flat patterns — `[` and `]` aren't in its `VALID_LAYER_SYMBOLS` — so it can't produce a grouped target and stays on the `final_norm` side.

Tests: `test_ungrouped_sharded_state_dict_keeps_hybrid_final_norm_key` and `test_sharded_state_dict_keeps_final_norm_key_without_transformer_keys`. The existing grouped assertions are unchanged; `test_group_sharded_state_dict_uses_logical_layer_keys` builds a `HybridStack` directly so it now passes `transformer_sharded_keys=True` itself.

[the remark this answers is not in the document]
Comment (Connor-XY) [reply megatron/core/models/hybrid/hybrid_block.py]: You're right that this was too broad, and it breaks more than old checkpoints. `docs/user-guide/hybrid-model-migration.md` documents `decoder.final_norm` as *the* HybridModel key, and both `tools/checkpoint/gpt_hybrid_conversion.py` and `tools/checkpoint/remap_gpt_dsa_to_mamba.py` rewrite GPT's `decoder.final_layernorm.*` into `decoder.final_norm.*` when converting. With the rename unconditional, every checkpoint those tools produce would have failed to load.

Now gated, in 08e141e: `HybridStack` takes a `transformer_sharded_keys` flag, and `HybridModel` sets it from whether the *full* layer pattern contains a bracketed group — full pattern rather than this rank's PP segment, so every stage agrees on the naming. Grouped patterns get `final_layernorm` (their logical layers are 1:1 with transformer layers, which is the whole point of the GPT cross-load); everything else keeps `final_norm`, so pre-existing hybrid checkpoints and converter output still load. The converter itself only accepts flat patterns — `[` and `]` aren't in its `VALID_LAYER_SYMBOLS` — so it can't produce a grouped target and stays on the `final_norm` side.

Tests: `test_ungrouped_sharded_state_dict_keeps_hybrid_final_norm_key` and `test_sharded_state_dict_keeps_final_norm_key_without_transformer_keys`. The existing grouped assertions are unchanged; `test_group_sharded_state_dict_uses_logical_layer_keys` builds a `HybridStack` directly so it now passes `transformer_sharded_keys=True` itself.

**Comment by Connor-XY** [comment]
Both strict-review findings addressed in ffcae6d + 08e141e, with replies inline.

1. `HybridModel._postprocess` now implements the `output_processor` early-return branch instead of accepting and dropping the argument.
2. The `final_norm` → `final_layernorm` sharded-key rename is gated on a new `transformer_sharded_keys` flag, which `HybridModel` derives from whether the full layer pattern has a bracketed group. Non-grouped hybrid models keep `decoder.final_norm.*`, which is what existing checkpoints and `tools/checkpoint/gpt_hybrid_conversion.py` output use.

**Comment by Connor-XY** [comment]
Rebased onto current main (076f61f). Conflicts were with the MLA-in-HybridModel port (#4452) and the mamba prefix-caching kernels (#5866):

- `Symbols` — kept both `MLA = "+"` and the group brackets; `VALID_LAYERS` has all of them.
- The Attention + MLA/DSA exclusion check now runs against the flattened layer list (so bracket characters don't confuse membership) and includes `MLA`, in both `_validate_pattern` and `_parse_segment_layers`.
- `HybridStack.__init__` — main's `_fuse_mla_down_proj` call and the `LayerSymbols.MLA` build branch live alongside the group/physical-offset restructure. The MLA branch uses the same `layer_number` / `pp_layer_offset` the attention and DSA branches do.
- `test_group_pattern` needed `'+': 0` in the expected counts dict.

Verified on 1 GB200 (hsg): 169 passed, 3 failed. The three are `test_hybrid_model_with_custom_process_groups[8-1-1 / 1-1-8 / 2-1-4]`, which need an 8-rank world and fail the same way before the rebase. Main's MLA tests (`TestMLADownProjFusion`, `test_mla_layer_types`, `test_mixed_attention_and_mla_layer_types`, `test_mixed_dsa_and_mla`) all pass next to the grouped-pattern tests.

**Comment by Connor-XY** [comment]
/ok to test [85bc2d4](https://github.com/NVIDIA/Megatron-LM/pull/4942/commits/85bc2d4ce45cac737ad0e66c5e951445342341e8)

**Comment by guihong-nv** [comment]
I checked `85bc2d4` and found a few things we should fix. The first two look like the biggest blockers:

1. **Grouped overlap fails on the first forward pass.** We override `get_fp8_context()` here, but the scheduler now calls `get_low_precision_context()`. That falls back to calling `get_inner_quantization_context()` on the `HybridStack`, which doesn't have that method. This fails even in BF16. We need to override the method the scheduler actually calls. [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/model_chunk_schedule_plan.py#L100-L108)

2. **The shared-expert gradient hooks never get picked up.** `backward_dw()` clears `self.layer`, and then the schedule node calls `parameters()` to find the gradient hooks. By then it returns nothing. With FSDP and delayed wgrad, that means the shared-expert gradients miss their accumulation/reduction callback. Can we keep the parameter references around until those hooks have been collected? [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/fine_grained_callables.py#L44-L61)

3. **Some valid groups can't be saved correctly.** For example, `[MM]` is accepted, but both Mamba layers end up with the same checkpoint key, including `decoder.layers.0.mixer.in_proj.weight`. Checkpoint validation then sees overlapping shards and rejects the save. We either need distinct keys for these sublayers or an early check that rejects groups with colliding keys. [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/hybrid_block.py#L857-L885)

4. **The CP check misses what's inside a group.** With CP > 1, contiguous linear layout, and zigzag attention layout, the outer stack can say no conversion is needed even though an inner group needs one. The check passes, but overlap skips the inner stack's layout conversion. Once the first issue is fixed, attention would receive tokens in the wrong layout. We need to check the nested stacks too. [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/model_chunk_schedule_plan.py#L131-L139)

5. **MLA is missing from the overlap path.** `+` is a valid layer symbol, but it isn't included in either the attention forward branch or the wgrad registration. A plain `+E` pattern raises a `ValueError` on the first attention forward. Can we add MLA to both branches? [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/fine_grained_callables.py#L269-L273), [wgrad registration](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/fine_grained_callables.py#L357-L365)

6. **MLP-norm offloading calls an API that no longer exists.** With `mlp_norm` offloading enabled, the combine step calls `off_interface.group_commit()` and raises an `AttributeError`. The current interface uses the manager instance's `group_offload()`, so this needs to follow that path. [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/fine_grained_callables.py#L232-L235)

7. **Direct `HybridStack` callers lose their checkpoint offset.** An existing caller using `pp_layer_offset=5` now saves its first layer under `layers.0` instead of `layers.5`, because the new logical offset defaults to zero. That breaks compatibility for callers building stacks directly. For ungrouped stacks, the default should still follow the physical offset. [Code](https://github.com/NVIDIA/Megatron-LM/blob/85bc2d4ce45cac737ad0e66c5e951445342341e8/megatron/core/models/hybrid/hybrid_block.py#L860-L865)

I checked these with small Python probes using the PR methods and stubbed dependencies, plus a base/head comparison for the checkpoint offset. I haven't run the distributed GPU tests.

**Review by guihong-nv — APPROVED** [review]
